### DISCUSSION TITLE
How to create a `Service` type in @xstate/fsm when using Typestate unions?

### DISCUSSION THREAD

**Discussion by lukekarrys** [statelyai/xstate#1944] [Q&A] [answered]
This is a minimal example I pulled from a project I'm working on, so apologies as it's a bit convoluted.  It's similar to the [@xstate/fsm TypeScript example in the docs](https://xstate.js.org/docs/packages/xstate-fsm/#typescript) in that I'm using the union of the typestates so Typescript knows about `state.context` inside `state.matches()` blocks.

My problem is that with my `Context`, `Event`, and `State` types I'm also trying to create a `Service` type that I can use with the `service` returned from `useMachine`.

```ts
interface Context {
  foo: string;
  bar: string;
  count?: number;
}

type State =
  | {
      value: "idle";
      context: Context & {
        count: undefined;
      };
    }
  | {
      value: "ready";
      context: Context & {
        count: number;
      };
    };

const machine = createMachine<Context, Event, State>({ ... })

type Service = StateMachine.Service<Context, Event, State>
const useCustomServiceHook = (service: Service) => { ... }

const Component = () => {
  cont [state, send, service] = useMachine(machine)

  // typescript complains about service
  useCusomServiceHook(service)

  return <div />
}
```

I can pretty easily workaround the issue and I think I undestand why it's happening but I was curious if there was a better way to type this with Typescript, or if I'm doing something else wrong with what `@xstate/fsm` is expecting for its `StateMachine.Service` types.

Here's a Codesandbox that reproduces the Typescript error: https://codesandbox.io/s/confident-snowflake-7izkb?file=/src/index.tsx

And the full Typescript error for good measure:
```
Argument of type 'Service<Context, Event, { value: any; context: Context; }>' is not assignable to parameter of type 'Service'.
  Types of property 'subscribe' are incompatible.
    Type '(listener: StateListener<State<Context, Event, { value: any; context: Context; }>>) => { unsubscribe: () => void; }' is not assignable to type '(listener: StateListener<State<Context, Event, State>>) => { unsubscribe: () => void; }'.
      Types of parameters 'listener' and 'listener' are incompatible.
        Types of parameters 'state' and 'state' are incompatible.
          Type 'State<Context, Event, { value: any; context: Context; }>' is not assignable to type 'State<Context, Event, State>'.
            Types of property 'matches' are incompatible.
              Type '<TSV extends any>(value: TSV) => this is { value: any; context: Context; }' is not assignable to type '<TSV extends "idle" | "ready">(value: TSV) => this is ({ value: "idle"; context: Context & { count: undefined; }; } extends { value: TSV; } ? { value: TSV; } & { value: "idle"; context: Context & { ...; }; } : never) | ({ ...; } extends { ...; } ? { ...; } & { ...; } : never)'.
                Type predicate 'this is { value: any; context: Context; }' is not assignable to 'this is State'.
                  Type '{ value: any; context: Context; }' is not assignable to type 'State'.
                    Type '{ value: any; context: Context; }' is not assignable to type '{ value: "ready"; context: Context & { count: number; }; }'.ts(2345)
```

**Comment by lukekarrys** [comment]
So I realized I was looking at the `@xstate/react` type signatures in [`src/useService.ts`](https://github.com/davidkpiano/xstate/blob/10a35861a50bed8d209cd9f444a093fe65d12e17/packages/xstate-react/src/useService.ts) instead of [`src/fsm.ts`](https://github.com/davidkpiano/xstate/blob/10a35861a50bed8d209cd9f444a093fe65d12e17/packages/xstate-react/src/fsm.ts#L40).

The line I linked to shows that the `service` returned from `useMachine` has the signature `StateMachine.Service<TContext, TEvent>` (without TState). So if I remove `State` from the Service type I'm trying to create, Typescript no longer errors, which is good enough for me.

I'm sure there's a good explanation as to why `service` is typed that way, and I would be really curious to hear what it is! 😄

**Reply by Andarist** [reply]
This might be an error - but I'm also not super sure. The equivalent in the XState-related hook includes that so I would count this as an oversight. Would you like to prepare a PR fixing this?

**Reply by lukekarrys** [reply]
Yeah, definitely! I just played around with it locally and think I got a working test case too. I'll PR it today sometime.

**Reply by lukekarrys** [reply]
Here's the created PR: https://github.com/davidkpiano/xstate/pull/1947

**Comment by lukekarrys** [comment, marked as answer]
This is now fixed in `@xstate/react@1.3.1` (https://github.com/davidkpiano/xstate/pull/1947).
